### PR TITLE
Add fapi validations for dcr tests

### PR DESCRIPTION
**Pull request by SachiniSiriwardene**
Added tests for FAPI validations durong DCR and for the newly added attributes
PR - https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2173
      - https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2178

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/OAuthDCRMTestCase.java [hunk 1]
```diff
@@ -17,32 +17,34 @@
...
 import org.json.simple.JSONArray;
 import org.json.simple.JSONObject;
 import org.json.simple.JSONValue;
-import org.testng.annotations.AfterClass;
-import org.testng.annotations.BeforeClass;
-import org.testng.annotations.DataProvider;
-import org.testng.annotations.Factory;
-import org.testng.annotations.Test;
+import org.json.simple.parser.JSONParser;
+import org.testng.annotations.*;
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/OAuthDCRMTestCase.java:36]: Remove start import

File: product-scenarios/4-single-sign-on/4.1-sso-for-web-app/4.1.4-sso-with-oidc/4.1.4.1-dcr/src/test/java/org/wso2/identity/scenarios/sso/test/dcr/DCRTestCase.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ private static Object[][] dcrConfigProvider() throws Exception {
                         getRegisterRequestJSON("request2.json"), getUpdateRequestJSON("request2.json"), ADMIN_USERNAME,
                         ADMIN_PASSWORD, SUPER_TENANT_DOMAIN
                 }
-        };
+                };
```
Comment (janakamarasena) [inline product-scenarios/4-single-sign-on/4.1-sso-for-web-app/4.1.4-sso-with-oidc/4.1.4.1-dcr/src/test/java/org/wso2/identity/scenarios/sso/test/dcr/DCRTestCase.java:101]: Unreleated change

**Comment by janakamarasena** [comment]
Let's extract the FAPI related tests to a separate test case as it requires a server restart.

Tests requiring server restart should be placed below https://github.com/wso2/product-is/blob/master/modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml#L282-L292

this is done to optimise the test execution time.

File: product-scenarios/4-single-sign-on/4.1-sso-for-web-app/4.1.4-sso-with-oidc/4.1.4.1-dcr/src/test/java/org/wso2/identity/scenarios/sso/test/dcr/DCRTestCase.java (hunk 2 above)
Comment (SachiniSiriwardene) [reply product-scenarios/4-single-sign-on/4.1-sso-for-web-app/4.1.4-sso-with-oidc/4.1.4.1-dcr/src/test/java/org/wso2/identity/scenarios/sso/test/dcr/DCRTestCase.java:101]: fixed in https://github.com/wso2/product-is/pull/17047/commits/1cdda47d18ad6c1a03a59c44b735c38946dade4f

File: modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml [hunk 3]
```diff
@@ -369,4 +369,10 @@
             <class name="org.wso2.identity.integration.test.rest.api.server.organization.management.v1.OrganizationManagementFailureTest"/>
         </classes>
     </test>-->
+    </test>
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml:372]: Seems like an unwanted tag.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import org.wso2.identity.integration.common.utils.ISIntegrationTest;
+import org.wso2.identity.integration.test.oauth2.dcrm.api.util.DCRUtils;
+import org.wso2.identity.integration.test.oauth2.dcrm.api.util.OAuthDCRMConstants;
+import org.wso2.identity.integration.test.util.Utils;
+
+import java.io.File;
+
+import static org.testng.Assert.assertEquals;
+
+public class FAPIDCRValidationsTestCase extends ISIntegrationTest {
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:40]: Lets add a class comment here

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+package org.wso2.identity.integration.test.oauth2.dcrm.api;
+
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.HttpClients;
+import org.json.simple.JSONObject;
+import org.testng.annotations.*;
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:27]: remove star import

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @DataProvider(name = "dcrmConfigProvider")
+    public static Object[][] dcrmConfigProvider() {
+        return new Object[][]{{TestUserMode.SUPER_TENANT_ADMIN}, {TestUserMode.TENANT_ADMIN}};
+    }
+
+    @BeforeClass(alwaysRun = true)
+    public void testInit() throws Exception {
+        super.init();
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:66]: Formatting

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import org.json.simple.JSONObject;
+import org.json.simple.JSONValue;
+import org.json.simple.parser.JSONParser;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class DCRUtils {
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java:34]: Add a class comment. check other relevant places

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.commons.codec.binary.Base64;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.util.EntityUtils;
+import org.json.simple.JSONObject;
+import org.json.simple.JSONValue;
+import org.json.simple.parser.JSONParser;
+
+import java.io.*;
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java:29]: Remove star import

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        if (tenant.equals("carbon.super")) {
+            return OAuthDCRMConstants.DCR_ENDPOINT_HOST_PART + OAuthDCRMConstants.DCR_ENDPOINT_PATH_PART;
+        } else {
+            return OAuthDCRMConstants.DCR_ENDPOINT_HOST_PART + "/t/" + tenant + OAuthDCRMConstants
+                    .DCR_ENDPOINT_PATH_PART;
+        }
+    }
+    public static String getAuthzHeader(String username, String password) {
+        return "Basic " + Base64.encodeBase64String((username + ":" + password).getBytes()).trim();
+    }
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java:83]: formatting. lets check other places as well.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/OAuthDCRMConstants.java [hunk 10]
```diff
@@ -46,6 +46,23 @@ public class OAuthDCRMConstants {
...
+    public static final String TLS_SUBJECT_DN = "tls_client_auth_subject_dn";
+    public static final String IS_PUSH_AUTH = "require_pushed_authorization_requests";
+    public static final String IS_SIGNED_REQUEST_OBJECT = "require_signed_request_object";
+    public static final String IS_CERTIFICATE_BOUND_ACCESS_TOKEN = "tls_client_certificate_bound_access_tokens";
+    public static final String SUBJECT_TYPE = "subject_type";
+    public static final String REQUEST_OBJECT_ENCRYPTION_ALGORITHM = "request_object_encryption_alg";
+    public static final String REQUEST_OBJECT_ENCRYPTION_METHOD = "request_object_encryption_enc";
+    public static final String JWKS_URI = "jwks_uri";
+
+
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/OAuthDCRMConstants.java:65]: Remove additional new lines

File: modules/integration/tests-integration/tests-backend/src/test/resources/artifacts/IS/oauth/dcr-fapi-validation-enabled.toml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+file_name = "wso2carbon.jks"
+password = "wso2carbon"
+
+[oauth]
+dcr.enable_sector_identifier_validation=true
+dcr.ssa_jkws="https://localhost:9853/oauth2/jwks"
+dcr.enable_fapi_enforcement=true
+oidc.fapi.enable_validation=true
+oidc.fapi.allowed_client_authentication_methods = ["private_key_jwt", "tls_client_auth"]
+oidc.fapi.allowed_signature_algorithms = ["PS256", "ES256"]
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/resources/artifacts/IS/oauth/dcr-fapi-validation-enabled.toml:36]: Add new line end of file

File: product-scenarios/4-single-sign-on/4.1-sso-for-web-app/4.1.4-sso-with-oidc/4.1.4.1-dcr/src/test/java/org/wso2/identity/scenarios/sso/test/dcr/DCRTestCase.java (hunk 2 above)
Comment (janakamarasena) [reply product-scenarios/4-single-sign-on/4.1-sso-for-web-app/4.1.4-sso-with-oidc/4.1.4.1-dcr/src/test/java/org/wso2/identity/scenarios/sso/test/dcr/DCRTestCase.java:101]: seems not yet updated.

**Review by janakamarasena — DISMISSED** [review]

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        JSONObject registerRequestJSON = DCRUtils.getRegisterRequestJSON("request6.json");
+
+        request.addHeader(HttpHeaders.AUTHORIZATION, DCRUtils.getAuthzHeader(username, password));
+        request.addHeader(HttpHeaders.CONTENT_TYPE, OAuthDCRMConstants.CONTENT_TYPE);
+        StringEntity entity = new StringEntity(registerRequestJSON.toJSONString());
+        request.setEntity(entity);
+        ObjectMapper mapper = new ObjectMapper();
+
+        HttpResponse response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 201, "Service Provider " +
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:163]: The created application should be deleted.

Comment (janakamarasena) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:163]: Also the test is for error scenarios. why do we need to check if a creation is successful?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        // Check error scenarios for update request.
+        HttpPut updateRequest = new HttpPut(DCRUtils.getPath(tenant) + client_id);
+        request.addHeader(HttpHeaders.AUTHORIZATION, DCRUtils.getAuthzHeader(username, password));
+        request.addHeader(HttpHeaders.CONTENT_TYPE, OAuthDCRMConstants.CONTENT_TYPE);
+        entity = new StringEntity(requestJSON.toJSONString());
+        request.setEntity(entity);
+        response = client.execute(updateRequest);
+
+        assertEquals(response.getStatusLine().getStatusCode(), 400, "Service Provider " +
+                "has not been created successfully");
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:178]: Shouldn't the message be something like. `Service Provider should not be created successfully`

Comment (janakamarasena) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:178]: check other relevant places as well

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        request.setEntity(entity);
+        HttpResponse response = client.execute(request);
+
+        assertEquals(response.getStatusLine().getStatusCode(), 400, "Service Provider " +
+                "has not been created successfully");
+        JSONObject errorResponse = DCRUtils.getPayload(response);
+        assertEquals(errorResponse.get("error"), errorCode);
+        assertEquals(errorResponse.get("error_description"), errorMessage);
+    }
+    @Test(alwaysRun = true, groups = "wso2.is", priority = 11,
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:147]: add new line

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java (hunk 5 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java (hunk 6 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java (hunk 8 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java (hunk 9 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/DCRUtils.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/OAuthDCRMConstants.java (hunk 10 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/util/OAuthDCRMConstants.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/resources/artifacts/IS/oauth/dcr-fapi-validation-enabled.toml (hunk 11 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/resources/artifacts/IS/oauth/dcr-fapi-validation-enabled.toml]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java (hunk 12 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:163]: this is to check whether the validations work during the application update flow

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java (hunk 13 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java]: changed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java (hunk 14 above)
Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java]: added

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        assertEquals(errorResponse.get("error"), errorCode);
+        assertEquals(errorResponse.get("error_description"), errorMessage);
+
+        // Delete application.
+        HttpDelete deleteRequest = new HttpDelete(DCRUtils.getPath(tenant) + client_id);
+        request.addHeader(HttpHeaders.AUTHORIZATION, DCRUtils.getAuthzHeader(username, password));
+
+        HttpResponse deleteResponse = client.execute(deleteRequest);
+        assertEquals(deleteResponse.getStatusLine().getStatusCode(), 204, "Service provider " +
+                "deleted successfully");
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java:188]:
```suggestion
        assertEquals(deleteResponse.getStatusLine().getStatusCode(), 204, "Service provider " +
                "deletion failed");
```

Comment (SachiniSiriwardene) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth2/dcrm/api/FAPIDCRValidationsTestCase.java]: changed

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**